apps/aecore/src/aec_vm_chain.erl

```diff
@@ -0,0 +1,140 @@
+    CTree0    = aec_trees:contracts(Trees),
+    Contract0 = aect_state_tree:get_contract(Key, CTree0),
+    Contract1 = aect_contracts:set_nonce(Nonce, Contract0),
+    CTree1    = aect_state_tree:enter_contract(Contract1, CTree0),
+    aec_trees:set_contracts(Trees, CTree1).
+
+
+%% @doc Get the balance of the contract account.
+-spec get_balance(chain_state()) -> non_neg_integer().
+get_balance(#state{ trees = Trees, account = PubKey }) ->
+    do_get_balance(PubKey, Trees).
+
+%% @doc Spend money from the contract account.
+-spec spend(pubkey(), non_neg_integer(), chain_state()) ->
+          {ok, chain_state()} | {error, term()}.
+spend(Recipient, Amount, State = #state{ trees   = Trees,
+                                         height  = Height,
+                                         account = ContractKey }) ->
+    case do_spend(Recipient, ContractKey, Amount, Trees, Height) of
+        {ok, Trees1}     -> {ok, State#state{ trees = Trees1 }};
+        Err = {error, _} -> Err
+    end.
+
+%% @doc Call another contract.
+-spec call_contract(pubkey(), non_neg_integer(), non_neg_integer(), binary(),
+                    [non_neg_integer()], chain_state()) ->
+        {ok, aec_vm_chain_api:call_result(), chain_state()} | {error, term()}.
+call_contract(Target, Gas, Value, CallData, CallStack,
+              State = #state{ trees   = Trees,
+                              height  = Height,
+                              account = ContractKey,
+                              nonce   = Nonce }) ->
+    VmVersion = 0,  %% TODO
+    {ok, CallTx} =
+        aect_call_tx:new(#{ caller     => ContractKey,
+                            nonce      => Nonce,
+                            contract   => Target,
+                            vm_version => VmVersion,
+                            fee        => 0,
+                            amount     => Value,
+                            gas        => Gas,
+                            gas_price  => 0,
+                            call_data  => CallData,
+                            call_stack => CallStack }),
+    case aetx:check_from_contract(CallTx, Trees, Height) of
+        Err = {error, _} -> Err;
+        {ok, Trees1} ->
+            {ok, Trees2} = aetx:process_from_contract(CallTx, Trees1, Height),
+            CallId  = aect_call:id(ContractKey, Nonce, Target),
+            Call    = aect_state_tree:get_call(Target, CallId, aec_trees:contracts(Trees2)),
+            GasUsed = aect_call:gas_used(Call),
+            Result  = case aect_call:return_value(Call) of
+                          %% TODO: currently we don't set any sensible return value on exceptions
+                          <<>> -> aec_vm_chain_api:call_exception(out_of_gas, GasUsed);
+                          Bin when is_binary(Bin) ->
+                            aec_vm_chain_api:call_result(Bin, GasUsed)
+                      end,
+            {ok, Result, State#state{ trees = Trees2, nonce = Nonce + 1 }}
```

Comment:
Why isn't the contract nonce increased in the trees rather than doing it in the `get_trees/1` function?

And I don't really follow. For which contract is the nonce increased? It looks to me as if the calling contract gets its nonce increased for every contract that it calls. Is this right?

Perhaps I just don't understand the nonce concept for contracts.

Reply:
After spending some more time on this, I understand why this is done the way it is. Looks good!